### ISSUE TITLE
site["scale"] is double counted at non-reparameterized sites

### ISSUE THREAD

**Issue by fritzo** [uber/pyro#531]
## The Problem

Consider the surrogate elbo at a nonreparameterized site (as [computed in](https://github.com/uber/pyro/blob/ece8ccc167695f81aa7ec1fed74a70c8fc15254b/pyro/infer/trace_elbo.py#L150) `Trace_ELBO`)
```py
log_p = model_site["log_pdf"]
log_q = guide_site["log_pdf"]
surrogate_elbo = log_p + (log_p - log_q).detach() * log_q
```
Now suppose this site occurs in an `irange` where we're subsampling 1% of the data, so that a `ScalePoutine` sets `site["scale"] = 100`. Then `Trace.log_pdf()` [will scale](https://github.com/uber/pyro/blob/ece8ccc167695f81aa7ec1fed74a70c8fc15254b/pyro/poutine/trace.py#L71) both `log_p` and `log_q` by 100. Therefore the `surrogate_elbo`'s first term `log_p` will be scaled by 100, but the second term `(log_p - log_q).detach() * log_q` will be scaled by 100 * 100.

## Scope

This should only affect models in one of the following scenarios:
1. The model has both global and local sites, and the local sites are non-reparameterized and subsampled.
2. The model has local sites that are non-reparameterized and subsampled and includes model parameters at that site (not only guide parameters).

## Possible Solutions

1. Handle `site["scale"]` separately, outside of `Trace.log_pdf()`. This scale is already wired through `Trace_ELBO` and `TraceGraph_ELBO` as `weight`.
  **Pros:** Avoids being too clever. Plays well with scale that varies across a batch (as in `BranchPoutine`).
  **Cons:** May require special treatment of `site["scale"]` in many places outside of `Trace_ELBO` and `TraceGraph_ELBO`.
2. Keep the existing logic of `Trace.log_pdf()`, and remove a factor of `site["scale"]` from the `(log_p - log_q).detach() * log_q` term.
  **Pros:** Minimally invasive.
  **Cons:** Seems like a hack. Does not play well with scales that vary across a batch.

**Comment by null-a** [comment]
> but the second term (log_p - log_q).detach() * log_q will be scaled by 100 * 100.

Is this really a bug?

`log_p_scaled` is a good estimate of `log_p_true`, similarly for `log_q`. If you want to compute an estimate of `(log_p_true - log_q_true) * log_q_true` then `(log_p_scaled - log_q_scaled) * log_q_scaled` gives you that.

If you get lucky and `log_p_scaled = log_p_true` (likewise for `q`) then you'll compute the exact thing. This wouldn't be the case if you pulled an extra scaling factor as proposed here?

**Comment by fritzo** [comment]
@null-a Yes I think you're right that this works as intended in `Trace_ELBO`. Thanks for explaining!

I'm still a little confused about `TraceGraph_ELBO`, which for an entirely local model [computes](https://github.com/uber/pyro/blob/ece8ccc167695f81aa7ec1fed74a70c8fc15254b/pyro/infer/tracegraph_elbo.py#L276) `surrogate_elbo` independently for each datum in a minibatch. I would expect the `surrogate_elbo` values to be scaled by `scale = size/subsample_size`, whereas in practice it is the `log_p` and `log_q` that are scaled, so that `surrogate_elbo` is scaled *twice*. Do you think it's possible that `Trace_ELBO` is correct, but `TraceGraph_ELBO` is incorrect?

**Comment by martinjankowiak** [comment]
i'm afraid @fritzo is right. there's an implicit rao-blackwellization that happens when you subsample which makes it a bit confusing. we'll fix it.
